### PR TITLE
refactor abstract operator into discreteoperator

### PR DESCRIPTION
**Pull request by gpescia**
This PR refactors most functionality away from AbstractOperator into a new DiscreteOperator.
This contains the first commit to #823 , and @PhilipVinc (who is writing here) thinks it's best we merge this ASAP so to avoid more merge conflicts..

### CODE REVIEW

File: netket/operator/_abstract_operator.py [hunk 1]
```diff
@@ -124,221 +107,11 @@ def conjugate(self, *, concrete=False) -> "AbstractOperator":
...
-            numba4jax won't work.
-            """
-        )
-
-    def __repr__(self):
-        return f"{type(self).__name__}(hilbert={self.hilbert})"
-
     def __matmul__(self, other):
-        if isinstance(other, np.ndarray) or isinstance(other, jnp.ndarray):
+        if isinstance(other, jnp.ndarray) or isinstance(other, jnp.ndarray):
```
Comment (PhilipVinc) [inline netket/operator/_abstract_operator.py:114]:
```suggestion
        if isinstance(other, np.ndarray) or isinstance(other, jnp.ndarray):
```

File: netket/operator/_discrete_operator.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+from netket.hilbert import AbstractHilbert
+from netket.operator import AbstractOperator
+
+
+class DiscreteOperator(AbstractOperator):
+    r"""This class is the base class for operators defined on a
+    discrete Hilbert space. Users interested in implementing new
+    quantum Operators for discrete Hilbert spaces should derive
+    their own class from this class"""
```
Comment (PhilipVinc) [inline netket/operator/_discrete_operator.py:17]:
```suggestion
    their own class from this class
    """
```

**Comment by PhilipVinc** [comment]
In essence, with this PR `AbstractOperator` becomes an empty shell that only has a field called `hilbert`.
All other functionality is moved to `DiscreteOperator` which maybe we should call `AbstractDiscreteOperator` (what do you think, @femtobit ?)

-- 

@gpescia I think this PR requires that most (all?) instances of `AbstractOperator` in `mcstate_expect.py` and `mcstate_expect_grad.py` used in dispatch be substituted with `DiscreteOperator`

**Comment by femtobit** [comment]
>  All other functionality is moved to DiscreteOperator which maybe we should call AbstractDiscreteOperator (what do you think, @femtobit ?)

`AbstractDiscreteOperator` is a bit verbose for my taste, I'd leave the name like it is now.

Doesn't `AbstractOperator` keep some `__matmul__` code here too? So its not a completely empty shell.

**Comment by PhilipVinc** [comment]
@femtobit I think that was a wrong judgement on my part.

I asked Gabriel to change it and remove all `__matmul__`code from `AbstractOperator` and move it to Discerete.
Basically `apply(x)` is not defined for the moment on non-discrete operators and it is not trivial on how to do it, and also `Squared` only works for discrete ATM so I think that's the best option.

Once more stuff will work with nondiscrete operators then maybe we can move this stuff back to AbstractOperator...

**Comment by PhilipVinc** [comment]
Last two things missing for me: Document this by adding DiscreteOperator to the API docs below AbstractOperator https://github.com/netket/netket/blob/master/docs/docs/api.rst#operators

**Review by PhilipVinc — APPROVED** [review]

**Comment by PhilipVinc** [comment]
Provided tests pass , this is good for me

**Review by femtobit — APPROVED** [review]

**Pull request merged**